File: leetcode_problems/python/lc94_binary_tree_inorder_traversal.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def preorder(
        self,
        node: TreeNode | None,
        mem: list[int],
    ):
        if node is None:
            return

        self.preorder(node.left, mem)
        mem.append(node.val)
        self.preorder(node.right, mem)

    def inorderTraversal(
        self,
        root: TreeNode | None,
    ) -> list[int]:
        mem = []
        self.preorder(root, mem)
        return mem


if __name__ == "__main__":
    root = TreeNode(1)
    root.right = TreeNode(2)
    root.right.left = TreeNode(3)

    data = Solution().inorderTraversal(root)
    print(data)

File: leetcode_problems/cpp/lc94_binary_tree_inorder_traversal.cpp
#include <vector>
using std::vector;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    void traverse(TreeNode *node, vector<int> &res)
    {
        if (node == nullptr)
        {
            return;
        }

        traverse(node->left, res);
        res.push_back(node->val);
        traverse(node->right, res);
    }

    vector<int> inorderTraversal(TreeNode *root)
    {
        vector<int> res;
        traverse(root, res);
        return res;
    }
};
